### ISSUE TITLE
object-curly-newline enforce either multiline or single line

### ISSUE THREAD

**Issue by CareLuLu-Gabriel** [eslint/eslint#13184]
**What rule do you want to change?**
object-curly-newline

**Does this change cause the rule to produce more or fewer warnings?**
More warnings. If disabled by default, there won't be breaking changes.

**How will the change be implemented? (New option, new default behavior, etc.)?**
There would be a new attribute. It could be `multilineConsistent` or `eitherMultilineOrSingleLine`. 

**Please provide some example code that this change will affect:**
```json
{
  "rules": {
    "object-curly-newline": ["error", {
      "ObjectExpression": {
        "multiline": true,
        "minProperties": 4,
        "consistent": true,
        "eitherMultilineOrSingleLine": true, // New property
      },
      "ObjectPattern": {
        "multiline": true,
        "minProperties": 4,
        "consistent": true,
        "eitherMultilineOrSingleLine": true, // New property
      },
      "ImportDeclaration": {
        "multiline": true,
        "minProperties": 4,
        "consistent": true,
        "eitherMultilineOrSingleLine": true, // New property
      },
      "ExportDeclaration": {
        "multiline": true,
        "minProperties": 4,
        "consistent": true,
        "eitherMultilineOrSingleLine": true, // New property
      }
    }],
  }
}
```

Examples of **incorrect** code for this rule with the `{ "eitherMultilineOrSingleLine": true }` option:
```js
import {
 a, b, c, d,
} from 'some-lib';

const varDeclaration = {
  a: 1, b: 2, c: 3, d: 4,
};

const functionDeclaration = ({
  a, b, c, d,
}) => {
  // do something
};

export default {
  a, b, c, d,
};
```

Examples of **correct** code for this rule with the `{ "eitherMultilineOrSingleLine": true }` option:
```js
// all multiline
import {
 a,
 b,
 c,
 d,
} from 'some-lib';

// all single line
import { e, f, g } from 'other-lib';

// all multiline
const varDeclaration = {
  a: 1,
  b: 2,
  c: 3,
  d: 4,
}; 

// all single line
const varDeclarationInline = { a: 1, b: 2, c: 3 };

// all multiline
const functionDeclaration = ({
  a,
  b,
  c,
  d,
}) => {
  // do something
};

// all single line
const functionDeclarationInline = ({ a, b, c }) => {
  // do something
};

// all multiline
export default {
  a,
  b,
  c,
  d,
};

// all single line
export default { a, b, c };
```

**What does the rule currently do for this code?**
It allows multiple properties on one line when newline was used for `{`.

**What will the rule do after it's changed?**
If the `eitherMultilineOrSingleLine` is `true`, then the whole statement will need to be on the same line or each property must be followed by a new line.

**Are you willing to submit a pull request to implement this change?**
Yes.

**Comment by mdjermanovic** [comment]
Hi @CareLuLu-Gabriel, thanks for the enhancement proposal!

Can you please clarify how is this different from the existing `multiline` option? All provided incorrect and correct examples are already incorrect/correct with `{ "multiline": true }`.

**Comment by CareLuLu-Gabriel** [comment]
@mdjermanovic `{ "multiline": true }` never allows single lines unless combined with `minProperties`. I will just show variable declaration for simplification.

## multiline

```
{
  "multile": true
}
```

Not allowed:
```
// Single line statement
const varDeclaration = { a: 1 };
const varDeclaration = { a: 1, b: 2 };
const varDeclaration = { a: 1, b: 2, c: 3 };

// Single line property
const varDeclaration = {
  a: 1, b: 2,
};
const varDeclaration = {
  a: 1, b: 2, c: 3,
};

// Multiline property
// NA
```

Allowed:
```
// Single line statement
// NA

// Single line property
// NA

// Multiline property
const varDeclaration = {
  a: 1,
};
const varDeclaration = {
  a: 1,
  b: 2,
};
const varDeclaration = {
  a: 1,
  b: 2,
  c: 3,
};
```

## multiline + minProperties

```
{
  "multile": true,
  "minProperties": 3,
}
```

Not allowed:
```
// Single line statement
const varDeclaration = { a: 1, b: 2, c: 3 };

// Single line property
const varDeclaration = {
  a: 1, b: 2,
};
const varDeclaration = {
  a: 1, b: 2, c: 3,
};

// Multiline property
const varDeclaration = {
  a: 1,
};
const varDeclaration = {
  a: 1,
  b: 2,
};
```

Allowed:
```
// Single line statement
const varDeclaration = { a: 1 };
const varDeclaration = { a: 1, b: 2 };

// Single line property
// NA

// Multiline property
const varDeclaration = {
  a: 1,
  b: 2,
  c: 3,
};
```

## multiline + minProperties + consistent

I'm not sure here if it's a bug on consistent (allowing `single line property`) or if it should be another property.

```
{
  "multile": true,
  "minProperties": 3,
  "consistent": true
}
```

Not allowed:
```
// Single line statement
const varDeclaration = { a: 1, b: 2, c: 3 };

// Single line property
// NA

// Multiline property
// NA
```

Allowed:
```
// Single line statement
const varDeclaration = { a: 1 };
const varDeclaration = { a: 1, b: 2 };

// Single line property
const varDeclaration = {
  a: 1, b: 2,
};
const varDeclaration = {
  a: 1, b: 2, c: 3,
};

// Multiline property
const varDeclaration = {
  a: 1,
};
const varDeclaration = {
  a: 1,
  b: 2,
};
const varDeclaration = {
  a: 1,
  b: 2,
  c: 3,
};
```

## (Proposal) multiline + minProperties + consistent + eitherMultilineOrSingleLine

```
{
  "multile": true,
  "minProperties": 3,
  "consistent": true,
  "eitherMultilineOrSingleLine": true
}
```

Not allowed:
```
// Single line statement
const varDeclaration = { a: 1, b: 2, c: 3 };

// Single line property
const varDeclaration = {
  a: 1, b: 2,
};
const varDeclaration = {
  a: 1, b: 2, c: 3,
};

// Multiline property
// NA
```

Allowed:
```
// Single line statement
const varDeclaration = { a: 1 };
const varDeclaration = { a: 1, b: 2 };

// Single line property
// NA

// Multiline property
const varDeclaration = {
  a: 1,
};
const varDeclaration = {
  a: 1,
  b: 2,
};
const varDeclaration = {
  a: 1,
  b: 2,
  c: 3,
};
```

**Comment by mdjermanovic** [comment]
> `{ "multiline": true }` never allows single lines unless combined with `minProperties`. I will just show variable declaration for simplification.

`{ "multiline": true }` allows single lines, without `minProperties`:

```js
/* eslint object-curly-newline: ["error", { "multiline": true }] */

const varDeclaration1 = { a: 1 };
const varDeclaration2 = { a: 1, b: 2 };
const varDeclaration3 = { a: 1, b: 2, c: 3 };
```

Please see this [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG9iamVjdC1jdXJseS1uZXdsaW5lOiBbXCJlcnJvclwiLCB7IFwibXVsdGlsaW5lXCI6IHRydWUgfV0gKi9cblxuY29uc3QgdmFyRGVjbGFyYXRpb24xID0geyBhOiAxIH07XG5jb25zdCB2YXJEZWNsYXJhdGlvbjIgPSB7IGE6IDEsIGI6IDIgfTtcbmNvbnN0IHZhckRlY2xhcmF0aW9uMyA9IHsgYTogMSwgYjogMiwgYzogMyB9O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by CareLuLu-Gabriel** [comment]
It seems there is no way to allow only up to 3 items written as inline statement and at the same time to block inline property.

<img width="1168" alt="Screen Shot 2020-04-15 at 4 07 46 PM" src="https://user-images.githubusercontent.com/4266269/79397606-538c8880-7f33-11ea-91dd-1fadcc1f2b29.png">

Here is the [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG9iamVjdC1jdXJseS1uZXdsaW5lOiBbXCJlcnJvclwiLCB7IFwibXVsdGlsaW5lXCI6IHRydWUgfV0gKi9cblxuLy8gSSB3YW50IHRvIGJsb2NrIHRoaXM6XG5cbi8vIDQgaXRlbXMgaW4gYSByb3cgKGRvZXNuJ3Qgd29yaylcbmNvbnN0IHZhckRlY2xhcmF0aW9uMSA9IHsgYTogMSwgYjogMiwgYzogMywgZDogNCB9O1xuXG4vLyBBbnl0aGluZyBtb3JlIHRoYW4gMSBpdGVtIGluIGEgcm93IG1peGVkIHdpdGggbmV3bGluZXNcbmNvbnN0IHZhckRlY2xhcmF0aW9uMiA9IHtcbiAgYTogMSwgYjogMixcbn07XG5cbi8vIEkgd2FudCB0byBhbGxvdyB0aGlzOlxuXG4vLyBVcCB0byAzIGl0ZW1zIGluIGEgcm93LlxuY29uc3QgdmFyRGVjbGFyYXRpb24zID0geyBhOiAxLCBiOiAyLCBjOiAzIH07XG5cbi8vIEFueSBudW1iZXIgb2YgaXRlbXMgYXMgb25lIHBlciBsaW5lIChkb2Vzbid0IHdvcmsgZm9yIDEgaXRlbSlcbmNvbnN0IHZhckRlY2xhcmF0aW9uNiA9IHtcbiAgYTogMSxcbn07XG5jb25zdCB2YXJEZWNsYXJhdGlvbjcgPSB7XG4gIGE6IDEsXG4gIGI6IDIsXG59O1xuY29uc3QgdmFyRGVjbGFyYXRpb244ID0ge1xuICBhOiAxLFxuICBiOiAyLFxuICBjOiAzLFxufTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=) of what I'm trying to accomplish.

**Comment by mdjermanovic** [comment]
If I understand correctly, the key issue is how to disallow this:

```js
const foo = {
    a, b, c, d
};
```

and enforce newlines between properties in this case.

This looks to me more like a case for a new option in [object-property-newline](https://eslint.org/docs/rules/object-property-newline). Something similar to the existing `allowAllPropertiesOnSameLine`, but taking into account the braces as well.

**Comment by CareLuLu-Gabriel** [comment]
Yes, mainly that:

```
const foo = {
    a, b,
};
const foo = {
    a, b, c,
};
const foo = {
    a, b, c, d, // ...
};
```

I think it'd work well on `object-property-newline`. We could either change `allowAllPropertiesOnSameLine = true` to taking into account the braces or add a new option to force taking into account braces. What do you think?

**Comment by mdjermanovic** [comment]
Changing how the existing option works, in this case, looks like a big breaking change.

I'd support a proposal for a new option in `object-property-newline` :+1:

**Comment by gcmarques** [comment]
@mdjermanovic closed this proposal and opened a new one for `object-property-newline`. https://github.com/eslint/eslint/issues/13189
